openerp/osv/fields.py

```diff
@@ -1078,7 +1081,10 @@ def unlink_all():
             elif act[0] == 5:
                 unlink_all()
             elif act[0] == 6:
-                unlink_all()
+                keep_ids = set(act[2])
+                for act1 in [v for v in values if v[0] in (1, 4)]:
```

Comment:
This loop is very brittle, and does not consider the order of the x2m commands, so it will sometimes produce invalid effects, while trying to introduce an optimization.
Haven't checked, but the generation of the commands might take care of adding all existing IDs to the `(6,0,[...])`, even the ones that will be edited with subsequent `(1,)` commands, to avoid the need for this reconstruction entirely.